metadata/dashboard/dashboard_metadata.go

```diff
@@ -1074,16 +1121,133 @@ func (m *MetadataServer) getSourceDataIterator(name, variant string, limit int64
 	return primary.IterateSegment(limit)
 }
 
+type VariantDuck interface {
+	Name() string
+	Variant() string
+	Tags() metadata.Tags
+}
+
+type TagResult struct {
+	Name    string   `json:"name"`
+	Variant string   `json:"variant"`
+	Tags    []string `json:"tags"`
+}
+
+func GetTagResult(param VariantDuck) TagResult {
+	return TagResult{
+		Name:    param.Name(),
+		Variant: param.Variant(),
+		Tags:    param.Tags(),
+	}
+}
+
+func (m *MetadataServer) GetTagError(code int, err error, c *gin.Context, resourceType string) *FetchError {
+	fetchError := &FetchError{StatusCode: code, Type: resourceType}
+	m.logger.Errorw(fetchError.Error(), "Metadata error", err)
+	return fetchError
+}
+
+func (m *MetadataServer) SetFoundVariantJSON(foundVariant VariantDuck, err error, c *gin.Context, resourceType string) {
+	if err != nil {
+		fetchError := m.GetTagError(500, err, c, resourceType)
+		c.JSON(fetchError.StatusCode, fetchError.Error())
+	}
+	c.JSON(http.StatusOK, GetTagResult(foundVariant))
+}
+
+func (m *MetadataServer) GetTags(c *gin.Context) {
+	name := c.Param("resource")
+	resourceType := c.Param("type")
+	nameVariant := metadata.NameVariant{Name: name, Variant: "default"}
+	switch resourceType {
+	case "features":
+		foundVariant, err := m.client.GetFeatureVariant(context.Background(), nameVariant)
+		m.SetFoundVariantJSON(foundVariant, err, c, resourceType)
+	case "labels":
+		foundVariant, err := m.client.GetLabelVariant(context.Background(), nameVariant)
+		m.SetFoundVariantJSON(foundVariant, err, c, resourceType)
+	case "training-sets":
+		foundVariant, err := m.client.GetTrainingSetVariant(context.Background(), nameVariant)
+		m.SetFoundVariantJSON(foundVariant, err, c, resourceType)
+	case "sources":
+		foundVariant, err := m.client.GetSourceVariant(context.Background(), nameVariant)
+		m.SetFoundVariantJSON(foundVariant, err, c, resourceType)
+	case "entities":
+		foundVariant, err := m.client.GetEntity(context.Background(), name)
+		m.SetFoundVariantJSON(foundVariant, err, c, resourceType)
+	case "users":
+		foundVariant, err := m.client.GetUser(context.Background(), name)
+		m.SetFoundVariantJSON(foundVariant, err, c, resourceType)
+	case "models":
+		foundVariant, err := m.client.GetModel(context.Background(), name)
+		m.SetFoundVariantJSON(foundVariant, err, c, resourceType)
+	case "providers":
+		foundVariant, err := m.client.GetProvider(context.Background(), name)
+		m.SetFoundVariantJSON(foundVariant, err, c, resourceType)
+	}
+}
+
+type TagRequestBody struct {
+	Tags []string `json:"tags"`
+}
+
```

Comment:
will make PostTags generic next. kept it as sourceVariant for quicker iteration.

Reply:
made generic using a switch. exploring using interfaces/structs to cut down on the boilerplate but Go is being a diva.